File: backend/services/stock_service.py
from typing import List
from models.stock import StockSearchRequest, StockItem, StockSearchResponse
from repositories.stock_repository import StockRepository
from repositories.base_repository import QueryError


class StockService:
    """Service layer for stock operations."""

    def __init__(self):
        self.repository = StockRepository()

    def search_stock(self, request: StockSearchRequest) -> StockSearchResponse:
        """
        Search stock by activity and flexible criteria.

        Returns:
            StockSearchResponse with items matching any criterion within activity
        """
        try:
            # Request validation is handled by Pydantic model_validator

            # Query repository
            results = self.repository.search_by_activity(request)

            # Convert to StockItem objects
            items = [
                StockItem(
                    art_code=row["art_code"],
                    stk_lieu=row["stk_lieu"],
                    stk_nosu=row["stk_nosu"],
                    qua_code=row["qua_code"],
                    stk_qte=row["stk_qte"],
                )
                for row in results
            ]

            # Build response
            if len(items) == 0:
                message = "No stock items found matching criteria"
            else:
                message = f"Found {len(items)} item(s)"

            return StockSearchResponse(
                status="success",
                message=message,
                items=items
            )

        except ValueError as e:
            # Validation error
            return StockSearchResponse(
                status="error",
                message=f"Invalid search criteria: {str(e)}",
                items=[]
            )
        except QueryError as e:
            # Database error
            return StockSearchResponse(
                status="error",
                message=f"Database error: {str(e)}",
                items=[]
            )
        except Exception as e:
            # Unexpected error
            return StockSearchResponse(
                status="error",
                message=f"Unexpected error: {str(e)}",
                items=[]
            )

File: backend-go/services/stock_service.go
package services

import (
	"errors"
	"strings"

	"stock-api/models"
	"stock-api/repositories"
)

// StockService is the service layer for stock operations.
type StockService struct {
	Repo *repositories.StockRepository
}

// NewStockService creates a StockService backed by the given repository.
func NewStockService(repo *repositories.StockRepository) *StockService {
	return &StockService{Repo: repo}
}

// SearchStock validates the search criteria, queries the repository and maps
// the resulting rows to the response model.
//
// At least one of ArtCode, StkLieu or StkNosu must be a non-nil, non-empty
// string. A validation failure returns an error and an "error" response. A
// database failure returns the error together with an "error" response that
// carries an empty (non-nil) item slice.
func (s *StockService) SearchStock(req models.StockSearchRequest) (models.StockSearchResponse, error) {
	if !hasSearchCriterion(req) {
		err := errors.New("At least one of art_code, stk_lieu, or stk_nosu must be provided")
		return models.StockSearchResponse{
			Status:  "error",
			Message: err.Error(),
			Items:   []models.StockItem{},
		}, err
	}

	rows, err := s.Repo.SearchByActivity(req)
	if err != nil {
		return models.StockSearchResponse{
			Status:  "error",
			Message: err.Error(),
			Items:   []models.StockItem{},
		}, err
	}

	items := make([]models.StockItem, 0, len(rows))
	for _, row := range rows {
		items = append(items, models.StockItem{
			ArtCode: row.ArtCode,
			StkLieu: row.StkLieu,
			StkNosu: row.StkNosu,
			QuaCode: row.QuaCode,
			StkQte:  row.StkQte,
		})
	}

	return models.StockSearchResponse{
		Status: "success",
		Items:  items,
	}, nil
}

// hasSearchCriterion reports whether at least one optional search field is set
// to a non-empty value.
func hasSearchCriterion(req models.StockSearchRequest) bool {
	return isNonEmpty(req.ArtCode) || isNonEmpty(req.StkLieu) || isNonEmpty(req.StkNosu)
}

// isNonEmpty reports whether p is non-nil and not blank once trimmed.
func isNonEmpty(p *string) bool {
	return p != nil && strings.TrimSpace(*p) != ""
}
